File: apps/android/app/src/main/java/fr/juliette/thecode/vault/Vault.java
package fr.juliette.thecode.vault;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Date;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.TimeZone;

/**
 * Carnet de métadonnées.
 *
 * Le carnet ne contient jamais de mot de passe ni de clef maîtresse. Une fuite
 * révèle les sites et les identifiants, pas les mots de passe.
 *
 * Règles de fusion : shared/spec/vault-merge.md
 */
public final class Vault {

    private static final String TAG = "TheCode";
    private static final String FILENAME = "vault.json";
    public static final int SCHEMA = 1;

    public String updatedAt = nowIso();
    public final List<VaultEntry> entries = new ArrayList<>();

    /** Un désaccord que la fusion refuse de trancher toute seule. */
    public static final class Conflict {
        public final String kind;
        public final String entryId;
        public final String detail;

        Conflict(String kind, String entryId, String detail) {
            this.kind = kind;
            this.entryId = entryId;
            this.detail = detail;
        }
    }

    public static String nowIso() {
        SimpleDateFormat fmt = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.ROOT);
        fmt.setTimeZone(TimeZone.getTimeZone("UTC"));
        return fmt.format(new Date());
    }

    /** Entrées couvrant ce domaine. Plusieurs = plusieurs comptes sur le site. */
    @NonNull
    public List<VaultEntry> findAllByDomain(String domain) {
        List<VaultEntry> out = new ArrayList<>();
        for (VaultEntry e : entries) {
            if (!e.deleted && e.coversDomain(domain)) out.add(e);
        }
        return out;
    }

    @Nullable
    public VaultEntry findByDomain(String domain) {
        List<VaultEntry> all = findAllByDomain(domain);
        return all.isEmpty() ? null : all.get(0);
    }

    /**
     * Enregistre les réglages d'un site, en créant l'entrée si besoin.
     *
     * Même sémantique que le CLI : la recherche se fait par domaine, pas par
     * siteKey, sinon « google.fr » créerait un doublon d'une entrée qui couvre
     * déjà « google.com » et « google.fr ».
     *
     * Le siteKey n'est jamais réécrit : il produit le mot de passe, le modifier
     * en changerait un déjà en service. L'appelant compare le siteKey rendu au
     * site saisi pour prévenir quand les deux diffèrent.
     */
    public VaultEntry upsert(String site, int length,
                             boolean lower, boolean upper, boolean symbols, boolean numbers) {
        VaultEntry entry = findByDomain(site);

        if (entry == null) {
            entry = VaultEntry.create(site, null);
            entries.add(entry);
        } else if (!entry.coversDomain(site)) {
            // Atteint seulement si findByDomain change de critère ; garde le
            // domaine saisi rattaché à l'entrée dans tous les cas.
            entry.domains.add(site);
            java.util.Collections.sort(entry.domains);
        }

        entry.v = VaultEntry.VERSION;
        entry.length = length;
        entry.lower = lower;
        entry.upper = upper;
        entry.symbols = symbols;
        entry.numbers = numbers;
        entry.updatedAt = nowIso();
        return entry;
    }

    /**
     * Supprime une entrée en laissant une pierre tombale : {@code deleted} et
     * {@code updatedAt} à maintenant. Retirer l'entrée ferait revenir la copie
     * de l'autre appareil à la synchronisation (voir vault-merge.md).
     *
     * @return faux si l'entrée n'existe pas ou est déjà supprimée.
     */
    public boolean delete(String id) {
        return delete(id, nowIso());
    }

    boolean delete(String id, String now) {
        for (VaultEntry e : entries) {
            if (e.id.equals(id) && !e.deleted) {
                e.deleted = true;
                e.updatedAt = now;
                return true;
            }
        }
        return false;
    }

    /** Identifiant tel qu'il entre dans la dérivation : absent vaut vide. */
    @NonNull
    public static String loginOf(@Nullable String login) {
        return login == null ? "" : login;
    }

    /** L'entrée de ce compte sur ce site : même domaine et même identifiant. */
    @Nullable
    public VaultEntry findAccount(String domain, @Nullable String login) {
        String wanted = loginOf(login);
        for (VaultEntry e : findAllByDomain(domain)) {
            if (loginOf(e.login).equals(wanted)) return e;
        }
        return null;
    }

    /**
     * Enregistre un compte : les réglages de l'entrée de ce domaine et de cet
     * identifiant, créée en v2 si elle n'existe pas.
     *
     * L'identifiant entre dans la dérivation v2 : deux identifiants sur un
     * même site sont deux comptes, donc deux entrées. Comme pour
     * {@link #upsert(String, int, boolean, boolean, boolean, boolean)}, le
     * siteKey d'une entrée existante n'est jamais réécrit.
     */
    public VaultEntry upsertAccount(String site, @Nullable String login, int length,
                                    boolean lower, boolean upper,
                                    boolean symbols, boolean numbers) {
        VaultEntry entry = findAccount(site, login);
        if (entry == null) {
            entry = VaultEntry.create(site, null);
            String value = loginOf(login);
            entry.login = value.isEmpty() ? null : value;
            entries.add(entry);
        }

        entry.v = VaultEntry.VERSION;
        entry.length = length;
        entry.lower = lower;
        entry.upper = upper;
        entry.symbols = symbols;
        entry.numbers = numbers;
        entry.updatedAt = nowIso();
        return entry;
    }

    /**
     * Fusionne deux carnets. Commutative et idempotente : l'ordre de
     * synchronisation des appareils ne doit pas changer le résultat.
     */
    public static Vault merge(Vault left, Vault right, List<Conflict> conflicts) {
        Map<String, VaultEntry> byId = new LinkedHashMap<>();
        for (VaultEntry e : left.entries) byId.put(e.id, e);

        for (VaultEntry incoming : right.entries) {
            VaultEntry existing = byId.get(incoming.id);
            byId.put(incoming.id,
                    existing == null ? incoming : mergeEntry(existing, incoming, conflicts));
        }

        Vault merged = new Vault();
        merged.entries.addAll(byId.values());
        // Collections.sort plutot que List#sort, qui demande l'API 24.
        java.util.Collections.sort(merged.entries, (a, b) -> a.id.compareTo(b.id));
        findDuplicates(merged.entries, idsOf(left.entries), idsOf(right.entries), conflicts);

        String latest = left.updatedAt.compareTo(right.updatedAt) >= 0
                ? left.updatedAt : right.updatedAt;
        for (VaultEntry e : merged.entries) {
            if (e.updatedAt.compareTo(latest) > 0) latest = e.updatedAt;
        }
        merged.updatedAt = latest;
        return merged;
    }

    private static VaultEntry mergeEntry(VaultEntry left, VaultEntry right,
                                         List<Conflict> conflicts) {
        // Égalité d'horodatage : on départage sur la représentation canonique.
        // Départager sur la position ne serait pas commutatif — chaque appareil
        // garderait le sien — et l'id ne peut pas servir, les deux entrées
        // portent la même.
        VaultEntry winner;
        int byDate = left.updatedAt.compareTo(right.updatedAt);
        if (byDate != 0) {
            winner = byDate > 0 ? left : right;
        } else {
            winner = left.canonical().compareTo(right.canonical()) <= 0 ? left : right;
        }

        VaultEntry merged = new VaultEntry();
        merged.id = left.id;
        merged.label = winner.label;
        merged.login = winner.login;
        merged.length = winner.length;
        merged.lower = winner.lower;
        merged.upper = winner.upper;
        merged.symbols = winner.symbols;
        merged.numbers = winner.numbers;
        merged.v = winner.v;

        // siteKey produit le mot de passe : on ne choisit jamais à la place de
        // l'utilisateur. On garde celui de gauche et on signale.
        merged.siteKey = winner.siteKey;
        if (!left.siteKey.equals(right.siteKey)) {
            conflicts.add(new Conflict("sitekey-divergent", left.id,
                    left.siteKey + " vs " + right.siteKey));
            merged.siteKey = left.siteKey;
        }

        // Union : une addition de chaque côté ne doit pas en effacer une autre.
        List<String> domains = new ArrayList<>(left.domains);
        for (String d : right.domains) {
            if (!domains.contains(d)) domains.add(d);
        }
        java.util.Collections.sort(domains);
        merged.domains = domains;

        // Un compteur ne recule pas : une valeur haute signifie déjà renouvelé.
        int high = Math.max(left.counter, right.counter);
        int low = Math.min(left.counter, right.counter);
        merged.counter = high;
        if (high != low && winner.counter == low) {
            conflicts.add(new Conflict("counter-recul", left.id,
                    "le plus recent porte " + low + ", on garde " + high));
        }

        // Une suppression se propage, sinon l'autre carnet ressusciterait
        // l'entrée.
        merged.deleted = left.deleted || right.deleted;

        // Une entrée n'est créée qu'une fois : la date la plus ancienne est la
        // vraie. Un carnet antérieur au champ ne doit pas l'effacer.
        if (left.createdAt == null) merged.createdAt = right.createdAt;
        else if (right.createdAt == null) merged.createdAt = left.createdAt;
        else merged.createdAt = left.createdAt.compareTo(right.createdAt) <= 0
                    ? left.createdAt : right.createdAt;

        merged.updatedAt = byDate >= 0 ? left.updatedAt : right.updatedAt;
        return merged;
    }

    private static Set<String> idsOf(List<VaultEntry> entries) {
        Set<String> ids = new HashSet<>();
        for (VaultEntry e : entries) ids.add(e.id);
        return ids;
    }

    /**
     * Signale les doublons que la fusion rapproche : le même compte créé à part
     * sur deux appareils, donc sous deux id. Un doublon déjà présent d'un côté
     * l'a été quand il y est entré — le resignaler à chaque fusion serait du
     * bruit.
     */
    private static void findDuplicates(List<VaultEntry> entries, Set<String> leftIds,
                                       Set<String> rightIds, List<Conflict> conflicts) {
        List<VaultEntry> live = new ArrayList<>();
        for (VaultEntry e : entries) {
            if (!e.deleted) live.add(e);
        }
        for (int i = 0; i < live.size(); i++) {
            for (int j = i + 1; j < live.size(); j++) {
                VaultEntry a = live.get(i);
                VaultEntry b = live.get(j);
                boolean apart = (leftIds.contains(a.id) && !rightIds.contains(a.id)
                        && rightIds.contains(b.id) && !leftIds.contains(b.id))
                        || (rightIds.contains(a.id) && !leftIds.contains(a.id)
                        && leftIds.contains(b.id) && !rightIds.contains(b.id));
                if (!apart || !loginOf(a.login).equals(loginOf(b.login))) continue;
                boolean shared = false;
                for (String d : b.domains) {
                    if (a.coversDomain(d)) {
                        shared = true;
                        break;
                    }
                }
                // Entrées triées par id : a porte le plus petit.
                if (shared) conflicts.add(new Conflict("doublon", a.id, b.id));
            }
        }
    }

    /** Ce qui part au serveur, et ce qui reste sur l'appareil. */
    public static final class PushSelection {
        public final List<VaultEntry> push;
        public final List<VaultEntry> localOnly;

        PushSelection(List<VaultEntry> push, List<VaultEntry> localOnly) {
            this.push = push;
            this.localOnly = localOnly;
        }
    }

    /**
     * Ce qui part au serveur quand le compte a un plafond, et ce qui reste sur
     * l'appareil. Voir shared/spec/vault-sync.md, « Synchronisation partielle ».
     *
     * Ce qui est déjà sur le serveur part toujours, pierres tombales comprises :
     * une modification doit pouvoir partir. Les places libres vont aux autres
     * entrées, les plus anciennes d'abord. Une entrée jamais synchronisée puis
     * supprimée n'a rien à propager.
     */
    public static PushSelection selectForPush(Vault vault, Collection<String> remoteIds,
                                              int maxEntries) {
        Set<String> remote = new HashSet<>(remoteIds);
        List<VaultEntry> onServer = new ArrayList<>();
        List<VaultEntry> others = new ArrayList<>();
        int liveOnServer = 0;
        for (VaultEntry e : vault.entries) {
            if (remote.contains(e.id)) {
                onServer.add(e);
                if (!e.deleted) liveOnServer++;
            } else if (!e.deleted) {
                others.add(e);
            }
        }
        java.util.Collections.sort(others, (a, b) -> {
            String da = a.createdAt != null ? a.createdAt : a.updatedAt;
            String db = b.createdAt != null ? b.createdAt : b.updatedAt;
            int byDate = da.compareTo(db);
            return byDate != 0 ? byDate : a.id.compareTo(b.id);
        });

        int free = Math.max(0, Math.min(others.size(), maxEntries - liveOnServer));
        List<VaultEntry> push = new ArrayList<>(onServer);
        push.addAll(others.subList(0, free));
        List<VaultEntry> localOnly = new ArrayList<>(others.subList(free, others.size()));
        java.util.Collections.sort(push, (a, b) -> a.id.compareTo(b.id));
        java.util.Collections.sort(localOnly, (a, b) -> a.id.compareTo(b.id));
        return new PushSelection(push, localOnly);
    }

    public static Vault fromJson(String json) throws JSONException {
        JSONObject root = new JSONObject(json);
        if (root.getInt("schema") != SCHEMA) {
            throw new JSONException("Carnet en version " + root.getInt("schema")
                    + ", attendu " + SCHEMA);
        }
        Vault v = new Vault();
        v.updatedAt = root.optString("updatedAt", nowIso());
        JSONArray arr = root.getJSONArray("entries");
        for (int i = 0; i < arr.length(); i++) {
            VaultEntry entry = VaultEntry.fromJson(arr.getJSONObject(i));
            // Une entrée hors v2 est écartée sans erreur : refuser tout le
            // carnet pour elle ferait perdre les autres. Elle disparaît du
            // carnet local à la prochaine écriture.
            if (entry.isSupported()) v.entries.add(entry);
        }
        return v;
    }

    /** JSON compact, pour le transfert : l'indentation gonflerait le QR. */
    public String toCompactJson() throws JSONException {
        return buildJson().toString();
    }

    public String toJson() throws JSONException {
        return buildJson().toString(2);
    }

    private JSONObject buildJson() throws JSONException {
        JSONObject root = new JSONObject();
        root.put("schema", SCHEMA);
        root.put("updatedAt", updatedAt);
        JSONArray arr = new JSONArray();
        // Le carnet n'écrit que de la v2, quoi qu'on ait mis dans la liste.
        for (VaultEntry e : entries) {
            if (e.isSupported()) arr.put(e.toJson());
        }
        root.put("entries", arr);
        return root;
    }

    public static Vault load(Context context) {
        File file = new File(context.getFilesDir(), FILENAME);
        if (!file.isFile()) return new Vault();
        try (java.io.FileInputStream in = new java.io.FileInputStream(file)) {
            // FileInputStream plutot que java.nio.file.Files, qui demande l'API 26.
            java.io.ByteArrayOutputStream buf = new java.io.ByteArrayOutputStream();
            byte[] chunk = new byte[8192];
            int n;
            while ((n = in.read(chunk)) != -1) buf.write(chunk, 0, n);
            return fromJson(new String(buf.toByteArray(), StandardCharsets.UTF_8));
        } catch (IOException | JSONException e) {
            Log.e(TAG, "Carnet illisible, on repart d'un carnet vide", e);
            return new Vault();
        }
    }

    /**
     * Efface le carnet local (« mot de passe oublié »). Si la synchronisation
     * est active, il revient à la prochaine synchronisation.
     */
    public static void wipe(Context context) {
        File file = new File(context.getFilesDir(), FILENAME);
        if (file.exists() && !file.delete()) {
            Log.e(TAG, "Echec de l'effacement du carnet");
        }
    }

    public void save(Context context) {
        File file = new File(context.getFilesDir(), FILENAME);
        File tmp = new File(context.getFilesDir(), FILENAME + ".tmp");
        updatedAt = nowIso();
        try {
            // Écriture atomique : une interruption ne doit pas laisser un
            // carnet tronqué, qui ferait perdre toutes les entrées.
            try (java.io.FileOutputStream out = new java.io.FileOutputStream(tmp)) {
                out.write(toJson().getBytes(StandardCharsets.UTF_8));
            }
            if (!tmp.renameTo(file)) {
                Log.e(TAG, "Echec du remplacement du carnet");
            }
        } catch (IOException | JSONException e) {
            Log.e(TAG, "Echec de l'ecriture du carnet", e);
        }
    }
}

File: apps/extension/vault.js
/**
 * Carnet de metadonnees.
 *
 * Le carnet ne contient jamais de mot de passe ni de clef maitresse : seulement
 * de quoi rejouer une derivation. Une fuite revele les sites et les
 * identifiants, pas les mots de passe.
 *
 * Schema : shared/vault.schema.json
 * Regles de fusion : shared/spec/vault-merge.md
 */

const VAULT_SCHEMA = 1;
const VAULT_STORAGE_KEY = "vault";

const VAULT_DEFAULT_CHARSET = { lower: true, upper: true, symbols: true, numbers: true };
const VAULT_DEFAULT_LENGTH = 20;

/**
 * Seule version d'entree admise dans le carnet.
 *
 * La v1 ne subsiste qu'en generation ponctuelle, hors carnet : une entree
 * `v != 2` est ecartee a chaque lecture et ne peut pas etre ecrite.
 * Voir shared/spec/vault-merge.md, « Uniquement des entrées v2 ».
 */
const VAULT_ENTRY_VERSION = 2;

function isV2Entry(entry) {
  return Boolean(entry) && entry.v === VAULT_ENTRY_VERSION;
}

/**
 * Ne garde que les entrees v2, sans erreur : un carnet ancien ou venu d'un
 * autre appareil reste lisible, ses entrees v1 disparaissent simplement.
 */
function keepV2Entries(vault) {
  if (!vault || !Array.isArray(vault.entries)) return vault;
  return { ...vault, entries: vault.entries.filter(isV2Entry) };
}

function nowIso() {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

function emptyVault() {
  return { schema: VAULT_SCHEMA, updatedAt: nowIso(), entries: [] };
}

function newEntry(siteKey, options = {}) {
  // Pas d'option `v` : une entree nait toujours en v2.
  const { label, domains, login = "", length = VAULT_DEFAULT_LENGTH, charset } = options;
  const now = nowIso();
  return {
    id: crypto.randomUUID(),
    label: label || siteKey,
    siteKey,
    domains: [...new Set(domains && domains.length ? domains : [siteKey])].sort(),
    login,
    counter: 1,
    length,
    charset: { ...VAULT_DEFAULT_CHARSET, ...(charset || {}) },
    v: VAULT_ENTRY_VERSION,
    createdAt: now,
    updatedAt: now,
  };
}

/** Entrees couvrant ce domaine. Plusieurs = plusieurs comptes sur le site. */
function findAllByDomain(vault, domain) {
  const target = String(domain).toLowerCase();
  return (vault.entries || []).filter(
    (e) => !e.deleted && e.domains.some((d) => d.toLowerCase() === target),
  );
}

function findByDomain(vault, domain) {
  return findAllByDomain(vault, domain)[0] || null;
}

/** Longueur maximale d'un identifiant : il entre dans la derivation v2. */
const VAULT_LOGIN_MAX = 120;

/** L'entree d'un compte precis : meme domaine et meme identifiant. */
function findByDomainAndLogin(vault, domain, login = "") {
  return findAllByDomain(vault, domain).find((e) => (e.login || "") === login) || null;
}

/**
 * Enregistre un compte depuis la popup.
 *
 * Le compte est designe par domaine + identifiant : deux identifiants sur un
 * meme site sont deux entrees. Une entree existante ne voit changer que sa
 * longueur et son jeu de caracteres — jamais son siteKey, qui produit le mot
 * de passe. Sinon, une entree v2 nait avec cet identifiant.
 */
function upsertSiteEntry(vault, { domain, login = "", length, charset }) {
  const existing = findByDomainAndLogin(vault, domain, login);
  if (existing) {
    existing.length = length;
    existing.charset = { ...VAULT_DEFAULT_CHARSET, ...(charset || {}) };
    existing.updatedAt = nowIso();
    return { entry: existing, updated: true };
  }
  const entry = newEntry(domain, { domains: [domain], login, length, charset });
  vault.entries.push(entry);
  return { entry, updated: false };
}

/** Representation stable, pour departager sans dependre de l'ordre. */

/**
 * Forme canonique d'une entree, pour departager de facon deterministe.
 *
 * Compacte, cles triees a tous les niveaux, `deleted` faux retire. La forme est
 * fixee par shared/spec/vault-merge.md et non laissee a JSON.stringify : deux
 * appareils qui n'ecrivent pas la meme chaine designent un gagnant different et
 * ne convergent jamais.
 *
 * JSON.stringify avec un tableau de cles ne convient pas : il applique le
 * filtre a tous les niveaux, ce qui vidait `charset`.
 */
function canonicalJson(entry) {
  const stable = (value) => {
    if (Array.isArray(value)) return value.map(stable);
    if (value === null || typeof value !== "object") return value;
    const out = {};
    for (const key of Object.keys(value).sort()) {
      if (value[key] !== undefined) out[key] = stable(value[key]);
    }
    return out;
  };

  const normalised = { ...entry };
  if (!normalised.deleted) delete normalised.deleted;
  return JSON.stringify(stable(normalised));
}

function mergeEntry(left, right, conflicts) {
  // Egalite d'horodatage : on departage sur la representation canonique.
  // Departager sur la position ne serait pas commutatif — chaque appareil
  // garderait le sien — et l'id ne peut pas servir, les deux entrees portent
  // la meme.
  let winner;
  if (left.updatedAt !== right.updatedAt) {
    winner = left.updatedAt > right.updatedAt ? left : right;
  } else {
    winner = canonicalJson(left) <= canonicalJson(right) ? left : right;
  }

  const merged = { ...winner };

  // siteKey produit le mot de passe : on ne choisit jamais a la place de
  // l'utilisateur. On garde celui de gauche et on signale.
  if (left.siteKey !== right.siteKey) {
    conflicts.push({
      kind: "sitekey-divergent",
      entryId: left.id,
      detail: `${left.siteKey} vs ${right.siteKey}`,
    });
    merged.siteKey = left.siteKey;
  }

  // Union : une addition de chaque cote ne doit pas en effacer une autre.
  merged.domains = [...new Set([...(left.domains || []), ...(right.domains || [])])].sort();

  // Un compteur ne recule pas : une valeur haute signifie deja renouvele.
  const high = Math.max(left.counter, right.counter);
  const low = Math.min(left.counter, right.counter);
  merged.counter = high;
  if (high !== low && winner.counter === low) {
    conflicts.push({
      kind: "counter-recul",
      entryId: left.id,
      detail: `le plus recent porte ${low}, on garde ${high}`,
    });
  }

  // Une suppression se propage, sinon l'autre carnet ressusciterait l'entree.
  if (left.deleted || right.deleted) merged.deleted = true;

  // Une entree n'est creee qu'une fois : la date la plus ancienne est la
  // vraie. Un carnet anterieur au champ ne doit pas l'effacer.
  const created = [left.createdAt, right.createdAt].filter(Boolean).sort();
  if (created.length) merged.createdAt = created[0];
  else delete merged.createdAt;

  merged.updatedAt = left.updatedAt > right.updatedAt ? left.updatedAt : right.updatedAt;
  return merged;
}

/**
 * Signale les doublons que la fusion rapproche : le meme compte cree a part
 * sur deux appareils, donc sous deux id. Un doublon deja present d'un cote
 * l'a ete quand il y est entre — le resignaler a chaque fusion serait du bruit.
 */
function findDuplicates(entries, leftIds, rightIds, conflicts) {
  const live = entries.filter((e) => !e.deleted);
  for (let i = 0; i < live.length; i += 1) {
    for (let j = i + 1; j < live.length; j += 1) {
      const a = live[i];
      const b = live[j];
      const apart =
        (leftIds.has(a.id) && !rightIds.has(a.id) && rightIds.has(b.id) && !leftIds.has(b.id)) ||
        (rightIds.has(a.id) && !leftIds.has(a.id) && leftIds.has(b.id) && !rightIds.has(b.id));
      if (!apart || (a.login || "") !== (b.login || "")) continue;
      const domains = new Set((a.domains || []).map((d) => d.toLowerCase()));
      if (!(b.domains || []).some((d) => domains.has(d.toLowerCase()))) continue;
      conflicts.push({ kind: "doublon", entryId: a.id, detail: b.id });
    }
  }
}

/**
 * Fusionne deux carnets. Commutative et idempotente : l'ordre de
 * synchronisation des appareils ne doit pas changer le resultat.
 */
function mergeVaults(left, right) {
  const conflicts = [];
  // Filet : ce qui sort d'une fusion est ecrit, et rien de v1 ne doit l'etre.
  const leftEntries = (left.entries || []).filter(isV2Entry);
  const rightEntries = (right.entries || []).filter(isV2Entry);
  const byId = new Map(leftEntries.map((e) => [e.id, e]));

  for (const entry of rightEntries) {
    const existing = byId.get(entry.id);
    byId.set(entry.id, existing ? mergeEntry(existing, entry, conflicts) : { ...entry });
  }

  const entries = [...byId.values()].sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));
  findDuplicates(
    entries,
    new Set(leftEntries.map((e) => e.id)),
    new Set(rightEntries.map((e) => e.id)),
    conflicts,
  );
  const updatedAt = [
    left.updatedAt || "",
    right.updatedAt || "",
    ...entries.map((e) => e.updatedAt),
  ]
    .filter(Boolean)
    .sort()
    .pop();

  return { vault: { schema: VAULT_SCHEMA, updatedAt: updatedAt || nowIso(), entries }, conflicts };
}

/**
 * Ce qui part au serveur quand le compte a un plafond, et ce qui reste sur
 * l'appareil. Voir shared/spec/vault-sync.md, « Synchronisation partielle ».
 *
 * Ce qui est deja sur le serveur part toujours, pierres tombales comprises :
 * une modification doit pouvoir partir. Les places libres vont aux autres
 * entrees, les plus anciennes d'abord. Une entree jamais synchronisee puis
 * supprimee n'a rien a propager.
 */
function selectForPush(vault, remoteIds, maxEntries) {
  const remote = new Set(remoteIds);
  const entries = vault.entries || [];
  const byId = (a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0);
  const onServer = entries.filter((e) => remote.has(e.id));
  const others = entries
    .filter((e) => !remote.has(e.id) && !e.deleted)
    .sort((a, b) => {
      const da = a.createdAt || a.updatedAt;
      const db = b.createdAt || b.updatedAt;
      if (da !== db) return da < db ? -1 : 1;
      return byId(a, b);
    });

  const free = Math.max(0, maxEntries - onServer.filter((e) => !e.deleted).length);
  return {
    push: [...onServer, ...others.slice(0, free)].sort(byId),
    localOnly: others.slice(free).sort(byId),
  };
}

async function loadVault(storage) {
  if (!storage) return emptyVault();
  try {
    const stored = await storage.get([VAULT_STORAGE_KEY]);
    const vault = stored?.[VAULT_STORAGE_KEY];
    if (!vault) return emptyVault();
    if (vault.schema !== VAULT_SCHEMA) {
      console.error("TheCode: carnet en version", vault.schema, "attendu", VAULT_SCHEMA);
      return emptyVault();
    }
    return keepV2Entries(vault);
  } catch (e) {
    console.error("TheCode: echec de la lecture du carnet", e);
    return emptyVault();
  }
}

async function saveVault(storage, vault) {
  if (!storage) return vault;
  vault.updatedAt = nowIso();
  try {
    await storage.set({ [VAULT_STORAGE_KEY]: vault });
  } catch (e) {
    console.error("TheCode: echec de l'ecriture du carnet", e);
  }
  return vault;
}

if (typeof module !== "undefined") {
  module.exports = {
    VAULT_SCHEMA,
    VAULT_STORAGE_KEY,
    VAULT_DEFAULT_CHARSET,
    VAULT_DEFAULT_LENGTH,
    VAULT_ENTRY_VERSION,
    isV2Entry,
    keepV2Entries,
    emptyVault,
    newEntry,
    findByDomain,
    findAllByDomain,
    findByDomainAndLogin,
    upsertSiteEntry,
    VAULT_LOGIN_MAX,
    mergeVaults,
    selectForPush,
    loadVault,
    saveVault,
    // Interne, expose pour le test contre la fixture partagee : la forme
    // canonique doit etre identique dans les cinq implementations.
    __canonicalJson: canonicalJson,
  };
}
